src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -172,20 +172,27 @@ private int[] getFilterIdsArray(final LeafReaderContext context) throws IOExcept
         if (filterWeight == null) {
             return new int[0];
         }
-        final BitSet filteredDocsBitSet = getFilteredDocsBitSet(context, this.filterWeight);
-        final int[] filteredIds = new int[filteredDocsBitSet.cardinality()];
-        int filteredIdsIndex = 0;
-        int docId = 0;
-        while (docId < filteredDocsBitSet.length()) {
-            docId = filteredDocsBitSet.nextSetBit(docId);
-            if (docId == DocIdSetIterator.NO_MORE_DOCS || docId + 1 == DocIdSetIterator.NO_MORE_DOCS) {
-                break;
-            }
-            filteredIds[filteredIdsIndex] = docId;
-            filteredIdsIndex++;
-            docId++;
+        return bitSetToIntArray(getFilteredDocsBitSet(context, this.filterWeight));
+    }
+
+    private int[] getParentIdsArray(final LeafReaderContext context) throws IOException {
+        if (knnQuery.getParentsFilter() == null) {
+            return null;
         }
-        return filteredIds;
+        return bitSetToIntArray(knnQuery.getParentsFilter().getBitSet(context));
+    }
+
+    private int[] bitSetToIntArray(final BitSet bitSet) {
+        final int[] intArray = new int[bitSet.cardinality()];
+        final BitSetIterator bitSetIterator = new BitSetIterator(bitSet, bitSet.cardinality());
```

Comment:
compute `bitSet.cardinality()` once, as this operation iterates over the words and compute number of set bits so it will be efficient to compute it once and reuse it again and again.

Reply:
ack